### PR TITLE
⚠️ CONFLICT! Lineage pull request for: skeleton

### PR DESCRIPTION
**Pull request by cisagovbot**
# Lineage Pull Request: CONFLICT #

[Lineage] has created this pull request to incorporate new changes found in an
upstream repository:

Upstream repository: [`https://github.com/cisagov/skeleton-tf-module.git`](https://github.com/cisagov/skeleton-tf-module.git)
Remote branch: `HEAD`

Check the changes in this pull request to ensure they won't cause issues with
your project.

The `lineage/skeleton` branch has **one or more unresolved merge conflicts**
that you must resolve before merging this pull request!

## How to resolve the conflicts ##

1. Take ownership of this pull request by removing any other assignees.

1. Clone the repository locally, and reapply the merge:

    ```console
    git clone git@github.com:cisagov/s3-read-role-tf-module.git s3-read-role-tf-module
    cd s3-read-role-tf-module
    git remote add skeleton https://github.com/cisagov/skeleton-tf-module.git
    git remote set-url --push skeleton no_push
    git switch develop
    git checkout -b lineage/skeleton --track origin/develop
    git pull skeleton HEAD
    git status
    ```

1. Review the changes displayed by the `status` command.  Fix any conflicts and
   possibly incorrect auto-merges.

1. After resolving each of the conflicts, `add` your changes to the
   branch, `commit`, and `push` your changes:

    ```console
    git add README.md examples/basic_usage/README.md examples/basic_usage/main.tf examples/basic_usage/outputs.tf examples/basic_usage/variables.tf main.tf outputs.tf variables.tf 
    git commit
    git push --force --set-upstream origin lineage/skeleton
    ```

    Note that you may _append_ to the default merge commit message
    that git creates for you, but *please do not delete the existing
    content*.  It provides useful information about the merge that is
    being performed.

1. Wait for all the automated tests to pass.

1. Check the "Everything is cool" checkbox below:

    - [x] ✌️ The conflicts in this pull request have been resolved.

1. Mark this draft pull request "Ready for review".

------------

**Note:** *You are seeing this because one of this repository's maintainers has
configured [Lineage] to open pull requests.*

For more information:

🛠 [Lineage] configurations for this project are stored in `.github/lineage.yml`

📚 [Read more about Lineage][Lineage]

[//]: # (lineage:metadata:{&quot;slayed&quot;:true})
[Lineage]: https://github.com/cisagov/action-lineage/ "Lineage GitHub Action"

### CODE REVIEW

**Review by jsf9k — APPROVED** [review]

**Review by dav3r — APPROVED** [review]
♨️

File: README.md [hunk 1]
```diff
@@ -76,27 +76,41 @@ This meta-role requires a permission policy similar to the following:
...
+| iam\_usernames | The list of IAM usernames allowed to assume the role.  If not provided, defaults to allowing any user in the specified account(s).  Note that including "root" in this list will override any other usernames in the list. | `list(string)` | `["root"]` | no |
+| role\_description | The description to associate with the IAM role (as well as the corresponding policy) that allows read-only access to the specified object(s) in the specified S3 buckets.  Note that the first "%s" in this value will get replaced with the s3\_bucket variable and the second "%s" will get replaced with the entity\_name variable. | `string` | `"Allows read-only access to S3 bucket %s required for %s."` | no |
+| role\_name | The name to assign the IAM role (as well as the corresponding policy) that allows read-only access to the specified S3 buckets.  Note that the first "%s" in this value will get replaced with the s3\_bucket variable and the second "%s" will get replaced with the entity\_name variable. | `string` | `"%s-ReadOnly-%s"` | no |
+| role\_tags | Tags to apply to the IAM role that allows read-only access to the specified S3 buckets. | `map(string)` | `{}` | no |
```
Comment (mcdonnnj) [inline README.md:111]: Is this being used in the same way that the `tags` variable is used in other TF repositories? This usage seems contrary to the changes you started in the skeleton:
https://github.com/cisagov/s3-read-role-tf-module/blob/8339b8eef83c40d444b9062dd751f5efc055cc3e/role.tf#L24-L30

**Review by mcdonnnj — COMMENTED** [review]
@jsf9k I want to confirm the answer to this question before I approve.

File: README.md (hunk 1 above)
Comment (jsf9k) [reply README.md:111]: It's a slightly different use, since the `role_tags` are only applied to a specific resource.  I left this behavior in place because I thought changing it would interfere with the way certain resources are labeled for use by the test users in [cisagov/skeleton-ansible-role-with-test-user](https://github.com/cisagov/skeleton-ansible-role-with-test-user) or [cisagov/skeleton-packer](https://github.com/cisagov/skeleton-packer), for example.  Perhaps @dav3r could comment on this aspect.

Even if that is the case, it might make more sense to rename this input `additional_tags_for_role` or something similar, since whatever tags are included in the variable will now be applied _in addition to_ the default tags with which the provider is configured.  I think it makes sense to make this change in a separate PR, though, since it will break a lot of other repos.

TL;DR - I was afraid that changing `role_tags` would break the AWS test users so I left it in place.  I agree that the variable needs to be renamed, at a minimum.

Comment (mcdonnnj) [reply README.md:111]: I was skimming over [repos that use this module](https://github.com/search?q=org%3Acisagov+s3-read-role-tf-module&type=code) and it looks like it's being called in a way that the individual Ansible roles would resolve just by switching to using `default_tags` in their providers. In [cisagov/terraform-state-read-role-tf-module](https://github.com/cisagov/terraform-state-read-role-tf-module) it looks like `role_vars` is mainly being passed through to remove `Workspace` tags in calling Terraform projects.

Comment (dav3r) [reply README.md:111]: I like @jsf9k's idea of renaming `role_tags` to `additional_tags_for_role` to make things a bit clearer.  That will allow those tags to be applied to the role created by this repo, while the default tags (from the provider) will be applied to both the role and the policy (and any other taggable resources that are added in the future).

Comment (jsf9k) [reply README.md:111]: I realized in debugging `terraform apply` errors in cisagov/ansible-role-cobalt-strike#35 that Terraform treats the case where `tags` is equal to the provider's `default_tags` as an error.  This means that this module is in some cases borked as it stands.  See [here](https://github.com/cisagov/ansible-role-cobalt-strike/blob/f93638345a3dea3fc531071ecb78796f74147e75/terraform/bucket_roles.tf#L21-L33), for instance.

I'd like to go ahead and merge this PR, then create a new one to make the change I described [here](https://github.com/cisagov/s3-read-role-tf-module/pull/14#discussion_r651194471), which will resolve the borking.  Are you OK with that @dav3r and @mcdonnnj?

Comment (dav3r) [reply README.md:111]: > I'd like to go ahead and merge this PR, then create a new one to make the change I described [here](https://github.com/cisagov/s3-read-role-tf-module/pull/14#discussion_r651194471), which will resolve the borking. Are you OK with that @dav3r and @mcdonnnj?

I'm good with that plan.

Comment (mcdonnnj) [reply README.md:111]: > I'd like to go ahead and merge this PR, then create a new one to make the change I described here, which will resolve the borking. Are you OK with that @dav3r and @mcdonnnj?

Won't changing the name from `role_vars` to `additional_tags_for_role` result in the same borking? If the values are the same as in `default_tags` it won't matter what we name it right?

Comment (jsf9k) [reply README.md:111]: @mcdonnnj - yes, but I will only pass the _additional_ (non-`default-tags`) tags to `additional_tags_for_role`.  I will have to make PRs for all the repos that use this module.

I will also add a `tags_to_ignore_for_role` parameter and use the `lifecycle --> ignore_changes` foo to ignore changes to those particular tags.  That won't stop the `Workspace` tag, for example, from being applied to the role, but it will stop it from changing.  If you or @dav3r can think of a better way to handle this case then let me know.  An alternative is to add the `Workspace` manually everywhere else via `tags`, but I don't like that solution either; maybe you guys do though.

Comment (mcdonnnj) [reply README.md:111]: Ah, ok. The second half is the part I was missing. That makes sense as a combined change. As far as `Workspace` tag should the caller be the one to worry about it as done [here](https://github.com/cisagov/cool-images-assessment-images/blob/8c2474e6ecbd0a097778f809ec98e2094e03b7e5/read_terraform_state_role.tf#L23)?

Comment (jsf9k) [reply README.md:111]: > Ah, ok. The second half is the part I was missing. That makes sense as a combined change. As far as `Workspace` tag should the caller be the one to worry about it as done [here](https://github.com/cisagov/cool-images-assessment-images/blob/8c2474e6ecbd0a097778f809ec98e2094e03b7e5/read_terraform_state_role.tf#L23)?

If the tag they want to ignore is one of the additional tags then they could remove it themselves; but, if it's one of the provider's default tags then they wouldn't be able to remove it like that.

Comment (dav3r) [reply README.md:111]: @jsf9k - I cannot think of a better solution here than what you have come up with.  I say go for it.

Comment (mcdonnnj) [reply README.md:111]: Yeah, this seems to be the best solution we have on hand. We can always revisit if we figure out a better way later.

**Review by mcdonnnj — APPROVED** [review]
We have reached an understanding so let us proceed!
